### PR TITLE
riotboot_serial: enter bootloader mode by pin

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

The `riotboot_serial` bootloader did not have the feature yet that it remains in bootloader mode when a certain pin is pulled / a button is pressed.

This makes flashing much easier as we don't have to rely on the right timing anymore. 

### Testing procedure

On a board that has a button, flash `bootloaders/riotboot_serial`.

Now reset the board and press the bootloader button (BTN0). After a bit, flash any other application with `PROGRAMMER=riotboot_serial RESET=`. 

The programmer should be able to connect to the board. Without this, it would attempt to load the first image after a timeout (and fail).

### CODE REVIEW

**Comment by aabadie** [comment]
Would it make sense to have LED0_PIN (if available) to oscillate if in bootloader mode ? This behavior is similar to the arduino bootloader and is quite useful.

File: sys/riotboot/serial.c [hunk 1]
```diff
@@ -53,6 +55,20 @@ static inline void uart_write_byte(uart_t uart, uint8_t data)
     uart_write(uart, &data, 1);
 }
 
+static inline bool _boot_pin(void)
```
Comment (chrysn) [inline sys/riotboot/serial.c:58]: This (and the later GPIO init) is duplicating _bootloader_alternative_mode of `bootloaders/riotboot_dfu/main.c`. I don't care which version survives, but have a slight preference to only have one. (Especially as they are imlpemented slightly differently).

Comment (benpicco) [reply sys/riotboot/serial.c:58]: I mean it's just `gpio_init()` and `gpio_read()`, I don't think the duplication of those two lines is bad.
`riotboot_serial` has a sync phase, so it allows to press the button at any time before the timeout occurs, it's not required to have the button pressed on reset.

Comment (chrysn) [reply sys/riotboot/serial.c:58]: Fair enough. I guess the better place to deduplicate here is to make these preprocessor macros into something more init-and-get. (I just don't like how the evaluation logic of the pin inversion is done slightly differently in two places).

**Comment by fjmolinas** [comment]
@chrysn have your changes been addressed? Should we go ahead with this one?

**Comment by fjmolinas** [comment]
ping @chrysn

**Comment by aabadie** [comment]
Tested on a nucleo board and works as advertised. The LED is not oscillating but blinking, but that's fine. @chrysn are your concerns addressed ?

**Review by aabadie — APPROVED** [review]
ACK

**Pull request merged**